### DISCUSSION TITLE
Fedoraproject package

### DISCUSSION THREAD

**Discussion by LecrisUT** [keycloak/keycloak#17293] [Ideas]
I will copy the issue #17066 because it was closed without proper discussion.

>>> Any plans to provide upstream packages for keycloak?
>>>
>> If you are talking about publishing RPMs for Keycloak that's not something we're planning, and very likely never will do. Please use discussions rather than issues for questions in the future though.
>>
> Yes, I am referring to publishing an RPM to src.fedoraproject.org. I believe this would be valuable.
>
> Are there licensing or design issues that prohibit keycloak to be published like that? If we contribute with a working copr, would you reconsider it, or at least support the maintenance of the copr repo by adding a webhook to automatically trigger the builds on tags?

@stianst can you elaborate on why RPMs are out of scope, and if the team can support the maintenance of the package by adding a simple webhook to copr?

**Comment by stianst** [comment]
We have had very few requests for RPMs, maybe 5-10 people asking for it in the many years we've been doing Keycloak. As such we simply can't prioritise it.

In terms of community maintained RPMs that's a nice idea, but question is how do you trust such community maintained RPMs? I'm sure you have the best intentions here, but at the same time those RPMs could be compromised at some point. For this reason we simply can't support the maintenance of these, because at the same time we would be officially blessing the RPMs and sending a message to the wider community that we trust them, while at the same time we have no way of actually validating that.

**Reply by LecrisUT** [reply]
Note that I am referring to "FedoraProject" not only a copr project. These are reviewed by the Fedora team so you cannot inject random code. Also the build instructions for the spec files are very straightforward to read ([here](https://docs.fedoraproject.org/en-US/java-packaging-howto/packaging_maven_project/#packaging_maven_project) is an example), so they can be easily vetted.

Of course if there wasn't ever a RPM, nobody is asking, but as soon as it is available people will use it. It's also notable that with keycloak's recent change from xml configuration to the `.conf` file, it is much easier to maintain an appropriate package. We are still lacking a `systemd` file upstream, but now it should be able to be developed.

I would say having it maintained in a distro package it is much easier to automatically patch vulnerabilities as they do `dnf upgrade`. By comparison, they would need to register to the package to get announced the vulnerabilities and then figure out how to upgrade from multiple major versions ago.

For the keycloak upstream we only need a bit of support:
- Must have: When there are vulnerabilities discovered, please notify us
- Suggestion: Include a spec file in the repo that we maintain via PRs upstream, and link it to a copr repo that some of us co-maintain.
  - This way we can easily detect the build status of the newest releases and keep fedoraproject release up to date
  - It is easier for new contributors since they only need to clone the main repo and run the build locally
  - This way you can also validate internally
- Request: Help in debugging the builds. The builds should be straightforward and we can run any necessary tests, but we wouldn't be experts of keycloak, so at the very least initially, it would be helpful to get some feedback of why a build or test is failing, until everything is setup.

**Reply by alexpdp7** [reply]
Note that the commercial version of Keycloak (Red Hat Single Sign On) has RPMs. The sources for that are available, but I think they are not easy to rebuild (at least with my limited packaging knowledge).

Personally, I'm looking for an update to Ipsilon, which was in EL7 and had very nice integration with FreeIPA. I was considering installing Keycloak on K8S, because the Keycloak operator looks like the way to get updates easily, and the plans for the operator are very enticing.

However, as I was researching this, I rechecked Ipsilon again. Ipsilon seems to be well-supported on Fedora, but I'd rather deploy it in EL... and someone is working on packaging it for EPEL 8. So now I'm leaning towards Ipsilon. It's pretty nice because it installs and updates as an RPM, and it autoconfigures for FreeIPA. If I can get OpenID going... well, I'd prefer EL9, but EL8 will still be good until 2029.
